File: angular-app/src/app/services/task.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Task } from '../models/task';

@Injectable({
  providedIn: 'root'
})
export class TaskService {
  private baseUrl = 'http://localhost:8080/task';

  constructor(private http: HttpClient) { }

  getAllTasks(): Observable<Task[]> {
    return this.http.get<Task[]>(this.baseUrl);
  }

  // Fetch tasks by user ID
  getTasksByUser(userId: number): Observable<Task[]> {
    return this.http.get<Task[]>(`${this.baseUrl}?userId=${userId}`);
  }

  // Fetch completed tasks
  getCompletedTasksByDate(completionDate: string): Observable<Task[]> {
    return this.http.get<Task[]>(`${this.baseUrl}?completion=true&completionDate=${completionDate}`);
  }

  createTask(task: Task): Observable<Task> {
    return this.http.post<Task>(this.baseUrl, task);
  }

  deleteTask(id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${id}`);
  }

  updateTask(task: Task): Observable<Task> {
    const url = `${this.baseUrl}/${task.id}`;
    return this.http.put<Task>(url, task);
  }
}

File: src/main/java/com/ed0leo/bbfweb/TaskService.java
package com.ed0leo.bbfweb;

import java.util.List;

public interface TaskService {
    List<Task> getTasks();
    List<Task> getTasksByUser(int userId);
    List<Task> getCompletedTasksByDate(String completionDate);
    List<Task> getCompletedTasksByDateAndUser(String completionDate, int userId);
    Task addTask(Task task);
    void deleteTask(int id);
    Task getTaskById(int id);
    Task updateTask(Task task);
}
